In [1]:
import pandas as pd

### MUESTRA DE DATOS

# 100K DATASET

In [3]:

#100K
users_1M = pd.read_csv('BBDD_1M/users.dat', delimiter='::', engine='python', header=None, names=['userId', 'gender', 'age', 'occupation', 'zip'], encoding='ISO-8859-1')
movies_1M = pd.read_csv('BBDD_1M/movies.dat', delimiter='::', engine='python', header=None, names=['movieId', 'title', 'genres'], encoding='ISO-8859-1')
ratings_1M = pd.read_csv('BBDD_1M/ratings.dat', delimiter='::', engine='python', header=None, names=['userId', 'movieId', 'rating', 'timestamp'], encoding='ISO-8859-1')

In [4]:
print(movies_1M.head())
'''
movieId: ID de la película.
title: Título de la película, incluyendo el año.
genres: Lista de géneros separados por "|".
'''

   movieId                               title                        genres
0        1                    Toy Story (1995)   Animation|Children's|Comedy
1        2                      Jumanji (1995)  Adventure|Children's|Fantasy
2        3             Grumpier Old Men (1995)                Comedy|Romance
3        4            Waiting to Exhale (1995)                  Comedy|Drama
4        5  Father of the Bride Part II (1995)                        Comedy


'\nmovieId: ID de la película.\ntitle: Título de la película, incluyendo el año.\ngenres: Lista de géneros separados por "|".\n'

In [5]:
print(users_1M.head())
'''
userId: ID del usuario.
movieId: ID de la película.
rating: Calificación de la película en una escala de 0.5 a 5.0.
timestamp: Momento en que se hizo la calificación.
'''

   userId gender  age  occupation    zip
0       1      F    1          10  48067
1       2      M   56          16  70072
2       3      M   25          15  55117
3       4      M   45           7  02460
4       5      M   25          20  55455


'\nuserId: ID del usuario.\nmovieId: ID de la película.\nrating: Calificación de la película en una escala de 0.5 a 5.0.\ntimestamp: Momento en que se hizo la calificación.\n'

In [6]:
print(ratings_1M.head())
'''
userId: ID del usuario.
movieId: ID de la película.
tag: Etiqueta asignada.
timestamp: Momento en que se añadió la etiqueta.
'''

   userId  movieId  rating  timestamp
0       1     1193       5  978300760
1       1      661       3  978302109
2       1      914       3  978301968
3       1     3408       4  978300275
4       1     2355       5  978824291


'\nuserId: ID del usuario.\nmovieId: ID de la película.\ntag: Etiqueta asignada.\ntimestamp: Momento en que se añadió la etiqueta.\n'

### DATOS CONCRETOS

In [7]:
df_1M = ratings_1M.copy()
df_1M.head()

,userId,movieId,rating,timestamp
0,1,1193,5,978300760
1,1,661,3,978302109
2,1,914,3,978301968
3,1,3408,4,978300275
4,1,2355,5,978824291


Cuantos usuarios, peliculas y ratings tenemos?

Usuarios

In [8]:
df_1M['userId'].unique()

array([   1,    2,    3, ..., 6038, 6039, 6040], dtype=int64)

100k --> 610 usuarios en total

In [9]:
len(df_1M['userId'].unique())

6040

Cuantas películas tenemos?

In [10]:
df_1M['movieId'].unique()

array([1193,  661,  914, ..., 2845, 3607, 2909], dtype=int64)

9724 películas en total 

In [11]:
len(df_1M['movieId'].unique())

3706

In [12]:
df_1M['rating'].head()

0    5
1    3
2    3
3    4
4    5
Name: rating, dtype: int64

Las valoraciones van desde los 0.5 a 5

In [13]:
print("Min rating = "+str(df_1M['rating'].min()))
print("Max rating = "+str(df_1M['rating'].max()))

Min rating = 1
Max rating = 5


El usuario que mas ratings ha puesto ha valorado 2698 películas, el que menos 20

In [14]:
# Suponiendo que el DataFrame se llama 'ratings'
ratings_count_per_user = df_1M.groupby('userId')['rating'].count()

# Obtener el máximo y el mínimo de ratings por usuario
max_ratings = ratings_count_per_user.max()
min_ratings = ratings_count_per_user.min()

print(f"Máximo de ratings por usuario: {max_ratings}")
print(f"Mínimo de ratings por usuario: {min_ratings}")

Máximo de ratings por usuario: 2314
Mínimo de ratings por usuario: 20


100836 valoraciones distintas

In [15]:
len(df_1M['rating'])

1000209

Tenemos valores nulos?

In [16]:
df_1M['userId'].isnull().sum()



0

In [17]:
df_1M['rating'].isnull().sum()

0

In [18]:
df_1M['movieId'].isnull().sum()

0

No hay valores nulos

### Pequeño resumen (BBDD 100K)

Tenemos una base de datos con 6040 usuarios y 3706 películas, donde cada usuario ha votado como minimo 20 películas y el que mas 2314, con un total de 1000209 valoraciones, sin valores nulos y con ratings que van desde 1 a 5 de puntuación.

## VISUALIZACIÓN DE LA ESCASEZ DE DATOS

Creación de la matriz de ratings de los usuarios sobre cada película

In [22]:
indice=list(df_1M['userId'].unique())
columnas=list(df_1M['movieId'].unique())
indice=sorted(indice)
columnas=sorted(columnas)
 
escasez_1M=pd.pivot_table(data=df_1M,values='rating',index='userId',columns='movieId')

In [21]:
escasez_1M

movieId,1,2,3,4,5,6,7,8,9,10,...,3943,3944,3945,3946,3947,3948,3949,3950,3951,3952
userId,,,,,,,,,,,,,,,,,,,,,
1,5.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5,NaN,NaN,NaN,NaN,NaN,2.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
6036,NaN,NaN,NaN,2.0,NaN,3.0,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6037,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6038,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [23]:
# Contar el número total de NaN en la matriz de escasez
total_nan = escasez_1M.isna().sum().sum()

print(f"Número total de valores NaN: {total_nan}")


Número total de valores NaN: 21384031


En total tenemos 21,384,031 valores NULOS en nuestra matriz, generando una abrumadora cantidad que tenemos que arreglar para poder generar nuestra algoritmo de CF